File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/I._Повтор/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    sequence: str = sys.stdin.readline().rstrip()
    n: int = len(sequence)
    lcp: list[int] = [0] * n
    for i in range(1, n):
        k: int = lcp[i - 1]
        while k and sequence[k] != sequence[i]: k = lcp[k - 1]
        if sequence[k] == sequence[i]: k += 1
        lcp[i] = k
    sys.stdout.write('{}\n'.format([1, n // (n - lcp[-1])][n % (n - lcp[-1]) == 0]))


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/I._Повтор/solution.cpp
#include <iostream>
#include <string>
#include <vector>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    std::string sequence;
    std::cin >> sequence;
    int n = sequence.size();
    std::vector<int> lcp (n, 0);
    for (int i = 1; i < n; ++i) {
        int k = lcp[i - 1];
        while (k && sequence[k] != sequence[i]) k = lcp[k - 1];
        if (sequence[k] == sequence[i]) ++k;
        lcp[i] = k;
    }
    std::printf("%d\n", (n % (n - lcp[n - 1]) == 0? n / (n - lcp[n - 1]) : 1));
}


int main() {
    solution();
}